File: old/objects/Player.lua
require "globals"
local love = require "love"

local Laser = require "objects.Laser"

function Player(lives)
    local SHIP_SIZE = 30
    local EXPLODE_DUR = 1
    local VIEW_ANGLE = math.rad(90)
    local LASER_LIMIT = 3
    local LASER_TIMER = 1.5
    local STARTING_LIVES = 3

    local statusEnum = {
        alive = 0,
        exploding = 1,
    }

    return {
        x = love.graphics.getWidth() / 2,
        y = love.graphics.getHeight() / 2,
        radius = SHIP_SIZE / 2,
        angle = VIEW_ANGLE,
        rotation = 0,
        lasers = {},
        thrusting = false,
        thrust = {
            x = 0,
            y = 0,
            speed = 5,
            big_flame = false,
            flame = 2.0
        },
        status = statusEnum.alive,
        explode_time = 0,
        lives = lives or STARTING_LIVES,

        isExploding = function(self)
            return self.status == statusEnum.exploding
        end,

        drawFlameThrust = function(self, fillType, color)
            love.graphics.setColor(color)

            local x, y, radius, flame = self.x, self.y, self.radius, self.thrust.flame
            local sin, cos = math.sin(self.angle), math.cos(self.angle)
            love.graphics.polygon(
                fillType,
                (x - (2/3 * cos + 0.5 * sin) * radius), (y + (2/3 * sin - 0.5 * cos) * radius),
                (x - radius * flame * cos),             (y + radius * flame * sin),
                (x - (2/3 * cos - 0.5 * sin) * radius), (y + (2/3 * sin + 0.5 * cos) * radius),
                (x - 1/2 * cos * radius),               (y + 1/2 * sin * radius)
            )
        end,

        drawShipShape = function (self, x, y, radius, angle, r, g, b, opacity)
            local sin, cos = math.sin(angle), math.cos(angle)

            love.graphics.setColor(r, g, b, opacity)
            love.graphics.polygon(
                "line",
                (x + 4/3 * cos * radius),         (y - 4/3 * sin * radius),
                (x - (2/3 * cos + sin) * radius), (y + (2/3 * sin - cos) * radius),
                (x - 1/2 * cos * radius),         (y + 1/2 * sin * radius),
                (x - (2/3 * cos - sin) * radius), (y + (2/3 * sin + cos) * radius)
            )
        end,

        draw = function(self, faded)
            local opacity = faded and 0.5 or 1

            if self.status == statusEnum.alive then
                if self.thrusting then
                    if not self.thrust.big_flame then
                        self.thrust.flame = self.thrust.flame - 10 * love.timer.getDelta()

                        if self.thrust.flame < 1.5 then
                            self.thrust.big_flame = true
                        end
                    else
                        self.thrust.flame = self.thrust.flame + 10 * love.timer.getDelta()

                        if self.thrust.flame > 2.5 then
                            self.thrust.big_flame = false
                        end
                    end

                    self:drawFlameThrust("fill", { 1, 0.7, 0.2 })
                    self:drawFlameThrust("line", { 1, 0.4, 0.1})
                end

                self:drawShipShape(self.x, self.y, self.radius, self.angle, 1, 1, 1, opacity)

                -- love.graphics.setColor(1, 1, 1, opacity)

                -- local x, y, radius = self.x, self.y, self.radius
                -- local sin, cos = math.sin(self.angle), math.cos(self.angle)
                -- love.graphics.polygon(
                --     "line",
                --     (x + 4/3 * cos * radius),         (y - 4/3 * sin * radius),
                --     (x - (2/3 * cos + sin) * radius), (y + (2/3 * sin - cos) * radius),
                --     (x - 1/2 * cos * radius),         (y + 1/2 * sin * radius),
                --     (x - (2/3 * cos - sin) * radius), (y + (2/3 * sin + cos) * radius)
                -- )
            elseif self.status == statusEnum.exploding then
                love.graphics.setColor(1, 0.2, 0.1, opacity)
                love.graphics.circle("fill", self.x, self.y, self.radius * 1.5)
                love.graphics.setColor(1, 0.5, 0.2, opacity)
                love.graphics.circle("fill", self.x, self.y, self.radius)
                love.graphics.setColor(1, 0.9, 0.2, opacity)
                love.graphics.circle("fill", self.x, self.y, self.radius * 0.5)
            end

            for _, laser in pairs(self.lasers) do
                laser:draw(faded)
            end

            if show_debugging then
                love.graphics.setColor(0, 1, 0, opacity)
                love.graphics.circle("fill", self.x, self.y, 1) --center
                love.graphics.circle("line", self.x, self.y, self.radius) --hitbox
            end
        end,

        drawLives = function(self, faded)
            local opacity = faded and 0.5 or 1
            local x_pos, x_offset, y_pos = 60, 20, 20
            if self.lives > 0 then
                for i = 1, self.lives - 1 do
                    self:drawShipShape(x_pos + x_offset * i, y_pos, self.radius / 2, VIEW_ANGLE, 0, 1, 0, opacity)
                end

                if self.status == statusEnum.exploding then
                    self:drawShipShape(x_pos + x_offset * self.lives, y_pos, self.radius / 2, VIEW_ANGLE, 0.8, 0, 0, opacity)
                elseif self.lives == 1 then
                    self:drawShipShape(x_pos + x_offset * self.lives, y_pos, self.radius / 2, VIEW_ANGLE, 0.8, 0.5, 0, opacity)
                else
                    self:drawShipShape(x_pos + x_offset * self.lives, y_pos, self.radius / 2, VIEW_ANGLE, 0, 1, 0, opacity)
                end
            end
        end,

        shoot = function(self)
            if #self.lasers < LASER_LIMIT and player.status == statusEnum.alive then
                table.insert(self.lasers, Laser(self.x, self.y, self.angle))
            end
        end,

        removeLaser = function(self, i)
            table.remove(self.lasers, i)
        end,

        move = function(self, dt)
            if self.status == statusEnum.alive then
                local friction = 0.7

                self.rotation = 2 * math.pi * dt

                if love.keyboard.isDown("a", "left", "kp4") then
                    self.angle = self.angle + self.rotation
                end

                if love.keyboard.isDown("d", "right", "kp6") then
                    self.angle = self.angle - self.rotation
                end

                if self.thrusting then
                    self.thrust.x = self.thrust.x + self.thrust.speed * math.cos(self.angle) * dt
                    self.thrust.y = self.thrust.y - self.thrust.speed * math.sin(self.angle) * dt
                elseif self.thrust.x ~= 0 or self.thrust.y ~= 0 then
                    self.thrust.x = self.thrust.x - friction * self.thrust.x * dt
                    self.thrust.y = self.thrust.y - friction * self.thrust.y * dt
                end

                self.x = self.x + self.thrust.x
                self.y = self.y + self.thrust.y

                if self.x + self.radius < 0 then
                    self.x = love.graphics.getWidth() + self.radius
                elseif self.x - self.radius > love.graphics.getWidth() then
                    self.x = -self.radius
                end

                if self.y + self.radius < 0 then
                    self.y = love.graphics.getHeight() + self.radius
                elseif self.y - self.radius > love.graphics.getHeight() then
                    self.y = -self.radius
                end
            elseif self.status == statusEnum.exploding then
                self.explode_time = self.explode_time + dt

                if self.explode_time > EXPLODE_DUR then
                    self:respawn()
                end
            end

            for i, laser in pairs(self.lasers) do
                laser:move(dt)

                if laser.travel_time > LASER_TIMER then
                    self:removeLaser(i)
                end

                if laser:doneExploding() then
                    self:removeLaser(i)
                end
            end
        end,

        respawn = function(self)
            self.lives = self.lives - 1
            if self.lives <= 0 then
                game:changeGameState("ended")
                return
            else
                self.status = statusEnum.alive
                self.explode_time = 0
                self.x = love.graphics.getWidth() / 2
                self.y = love.graphics.getHeight() / 2
            end
        end,

        explode = function(self)
            self.status = statusEnum.exploding
            self.thrust.x = 0
            self.thrust.y = 0
        end,
    }
end

return Player

File: src/objects/player.py
from typing import List
import pygame
from enum import Enum
from math import pi, radians, sin, cos
from components.sfx import Sfx, SoundMode
import components.sfx as s

from globals import *
from utils import *
from utils import draw_big_explosion, draw_hitbox, draw_polygon, draw_ship_shape
from objects.laser import Laser

class Player:
    SHIP_SIZE = 30.0
    EXPLODE_DUR = 1.0
    VIEW_ANGLE = radians(90)
    LASER_LIMIT = 5.0
    LASER_TIMER =  2.0
    STARTING_LIVES = 3
    RESPAWN_TIME = 3

    class Status(Enum):
        ALIVE = 0
        EXPLODING = 1
        RESPAWNING = 2

    def __init__(self, sfx: Sfx, lives = STARTING_LIVES):
        self.sfx = sfx
        self.x = WIN_WIDTH / 2.0
        self.y = WIN_HEIGHT / 2.0
        self.radius = self.SHIP_SIZE / 2.0
        self.angle = self.VIEW_ANGLE
        self.rotation = 0.0
        self.lasers : List[Laser] = []
        self.thrusting = False
        self.thrust_x = 0.0
        self.thrust_y = 0.0
        self.thrust_speed = 5.0
        self.thrust_big_flame = False
        self.thrust_flame = 2.0
        self.status = self.Status.RESPAWNING
        self.explode_time = 0
        self.lives = lives
        self.respawn_time = self.RESPAWN_TIME

    def draw_flame_thrust(self, win, fill_type, color):
        width = 2 if fill_type == "line" else 0

        x, y, r, flame = self.x, self.y, self.radius, self.thrust_flame
        si, co = sin(self.angle), cos(self.angle)

        points = [
            (x - (2/3 * co + 0.5 * si) * r, y + (2/3 * si - 0.5 * co) * r),
            (x - r * flame * co,           y + r * flame * si),
            (x - (2/3 * co - 0.5 * si) * r, y + (2/3 * si + 0.5 * co) * r),
            (x - 1/2 * co * r,              y + 1/2 * si * r)
        ]
        draw_polygon(win, color, points, width)

    def draw(self, win, dt, debugging):
        if self.status in [self.Status.ALIVE, self.Status.RESPAWNING]:
            if self.thrusting:

                #this blocks alternates between growing flame and shrinking flame
                #could probably be done a different way but that's what the tutorial went with
                if not self.thrust_big_flame:
                    self.thrust_flame -= 10 * dt

                    if self.thrust_flame < 1.5:
                        self.thrust_big_flame = True
                else:
                    self.thrust_flame += 10 * dt

                    if self.thrust_flame > 2.5:
                        self.thrust_big_flame = False
                
                self.draw_flame_thrust(win, "fill", THRUST_YELLOW)
                self.draw_flame_thrust(win, "line", THRUST_ORANGE)

            if debugging:
                draw_hitbox(win, GREEN, (self.x, self.y), self.radius, 1)

            color = WHITE
            if self.status == self.status.RESPAWNING:
                # this makes the player blink every 1/3 of a second between white and gray
                value = 128 if int(self.respawn_time * 10 % 6) > 2 else 255
                color = (value, value, value)

            draw_ship_shape(win, self.x, self.y, self.radius, self.angle, color, 2, debugging)
        elif self.status == self.status.EXPLODING:
            draw_big_explosion(win, EXPLOSION_RED, EXPLOSION_ORANGE, EXPLOSION_YELLOW, (self.x, self.y), self.radius * 1.5, self.radius, self.radius * 0.5)

        for laser in self.lasers:
            laser.draw(win)

    def draw_lives(self, win, debugging):
        x_pos, x_offset, y_pos = 60, 20, 25

        if self.lives > 0:
            color = GREEN
            for i in range(1, self.lives + 1):
                if i == self.lives:
                    if self.status == self.Status.EXPLODING:
                        color = HURT_RED
                    elif self.lives == 1:
                        color = WARNING_ORANGE
                draw_ship_shape(win, x_pos + x_offset * i, y_pos, self.radius / 2, self.VIEW_ANGLE, color, 1, debugging)

    def shoot(self):
        if self.status in [self.Status.RESPAWNING, self.Status.ALIVE] and len(self.lasers) < self.LASER_LIMIT:
            x, y = self.x + 4/3 * cos(self.angle) * self.radius, self.y - 4/3 * sin(self.angle) * self.radius
            self.lasers.append(Laser(x, y, self.angle))
            self.sfx.play_sfx(s.LASER)
    
    def make_invulnerable(self, time = RESPAWN_TIME):
        self.status = self.Status.RESPAWNING
        self.respawn_time = time

    def respawn(self, time = RESPAWN_TIME):
        self.lives = self.lives - 1
        if self.lives <= 0:
            return False
        else:
            self.make_invulnerable(time)
            self.explode_time = 0
            self.x = WIN_WIDTH / 2
            self.y = WIN_HEIGHT / 2
            self.angle = self.VIEW_ANGLE
            return True

    def move(self, dt):
        if self.status in [self.Status.ALIVE, self.Status.RESPAWNING]:
            friction = 0.7

            self.rotation = 2 * pi * dt

            keys = pygame.key.get_pressed()
            if keys[pygame.K_a] or keys[pygame.K_LEFT] or keys[pygame.K_KP4]:
                self.angle += self.rotation
            if keys[pygame.K_d] or keys[pygame.K_RIGHT] or keys[pygame.K_KP6]:
                self.angle -= self.rotation

            if self.thrusting:
                self.thrust_x += self.thrust_speed * cos(self.angle) * dt
                self.thrust_y -= self.thrust_speed * sin(self.angle) * dt
                self.sfx.play_sfx(s.THRUSTER, SoundMode.SLOW)
            elif self.thrust_x != 0 or self.thrust_y != 0:
                self.thrust_x -= friction * self.thrust_x * dt
                self.thrust_y -= friction * self.thrust_y * dt
                self.sfx.stop_sfx(s.THRUSTER)

            self.x = wrap_x(self.x + self.thrust_x, self.radius)
            self.y = wrap_y(self.y + self.thrust_y, self.radius)

            if self.status == self.Status.RESPAWNING:
                self.respawn_time -= dt
                if self.respawn_time <= 0:
                    self.respawn_time = 0
                    self.status = self.Status.ALIVE

        elif self.status == self.Status.EXPLODING:
            self.sfx.stop_sfx(s.THRUSTER)
            self.explode_time += dt

            if self.explode_time > self.EXPLODE_DUR:
                if not self.respawn():
                    return False
        
        toRemove = []
        for laser in self.lasers:
            if not laser.move(dt):
                toRemove.append(laser)
            elif laser.travel_time > self.LASER_TIMER:
                toRemove.append(laser)
            elif laser.status == laser.Status.DESTROYED:
                toRemove.append(laser)
        for laser in toRemove:
            self.lasers.remove(laser)

        return True
    
    def explode(self):
        self.sfx.play_sfx(s.SHIP_EXPLOSION)
        self.status = self.Status.EXPLODING
        self.thrust_x = 0
        self.thrust_y = 0
